lib/ecto/pools/ownership.ex

```diff
@@ -0,0 +1,202 @@
+      ]
+      supervise(children, strategy: :simple_one_for_one)
+    end
+
+    def new_owner(name) do
+      Supervisor.start_child(name, [])
+    end
+  end
+
+  defmodule Sup do
+    use Supervisor
+
+    def start_link(connection, opts) do
+      Supervisor.start_link(__MODULE__, {connection, opts})
+    end
+
+    def init({connection, opts}) do
+      {pool, opts} = Keyword.pop(opts, :ownership_pool)
+      name         = Keyword.fetch!(opts, :pool_name)
+      pool_name    = Module.concat(name, Inner)
+      pool_opts    = Keyword.put(opts, :pool_name, pool_name)
+      sup_name     = Module.concat(name, Elixir.ServerSup)
+
+      :ets.new(name, [:named_table, :public, read_concurrency: true])
+      :ets.insert(name, {:metadata, sup_name, pool_name, pool})
+
+      children = [
+        supervisor(ServerSup, [sup_name, pool, pool_name, name]),
+        supervisor(pool, [connection, pool_opts])
+      ]
+      supervise(children, strategy: :rest_for_one)
+    end
+  end
+
+  @behaviour Ecto.Pool
+  @timeout 5_000
+
+  def start_link(connection, opts) do
+    Sup.start_link(connection, opts)
+  end
+
+  def ownership_checkout(repo, strategy \\ nil, timeout \\ @timeout) do
+    {_, pool, _} = repo.__pool__
+
+    if :ets.member(pool, self) do
+      raise "process already owns a worker"
+    else
+      supervisor = :ets.lookup_element(pool, :metadata, 2)
+      {:ok, pid} = ServerSup.new_owner(supervisor)
+
+      case GenServer.call(pid, {:ownership_checkout, strategy, timeout}) do
+        {:ok, checkout} ->
+          unless :ets.insert_new(pool, {self, pid, checkout}) do
+            GenServer.call(pid, {:ownership_checkin, timeout}, timeout)
+            raise "race condition"
+          end
+        {:error, _} = error ->
+          error
+      end
+    end
+  end
+
+  def ownership_checkin(repo, timeout \\ @timeout) do
+    {_, pool, _, } = repo.__pool__
+    if :ets.member(pool, self) do
+      [{_, pid, _}] = :ets.lookup(pool, self)
+      GenServer.call(pid, {:ownership_checkin, timeout}, timeout)
+    else
+      raise "process doesn't own a worker"
+    end
+  end
+
+  def checkout(pool, _timeout) do
+    case :ets.lookup(pool, self) do
+      [{_, _, {worker, mod_conn, mode}}] ->
+        {:ok, worker, mod_conn, mode, 0}
+      [] ->
+        raise "..."
+    end
+  end
+
+  def checkin(_pool, _worker, _timeout) do
+    :ok
+  end
+
+  def break(pool, worker, timeout) do
+    if :ets.member(pool, self) do
+      [{:metadata, _supervisor, inner_pool, module}] = :ets.lookup(pool, :metadata)
+      module.break(inner_pool, worker, timeout)
+    else
+      raise "..."
+    end
+  end
+
+  def checkout_transaction(pool, timeout) do
+    if :ets.member(pool, self) do
+      [{:metadata, _supervisor, inner_pool, module}] = :ets.lookup(pool, :metadata)
+      [{_, _, {worker, mod_conn, mode}}] = :ets.lookup(pool, self)
+      module.open_transaction(inner_pool, worker, timeout)
```

Comment:
We need to serialize all pool operations through the worker. That's why we introduced the worker in the first place, so it serializes everything once we support open_transaction with allow/disallow. :)